### PR TITLE
Add experimental Prometheus metrics endpoints

### PR DESCRIPTION
**Pull request by fasaxc**
Todo:

- [x] Expose existing metrics
- [x] Add metrics around resync
- [x] Add config to enable.  Diasabled by default to avoid port binding clash.
- [x] Docs.
- [x] Package prometheus client library
  - [x] Ubuntu (already existed in Xenial, so used backport tool)
  - [x] RH (added to Jenkins)
- [x] UT

### CODE REVIEW

File: calico/felix/futils.py [hunk 1]
```diff
@@ -400,14 +402,26 @@ def register_diags(name, fn):
...
 
+def sanitize_name(name):
+    return re.sub(r'[^a-zA-Z0-9]', '_', name)
+
+
 class StatCounter(object):
     def __init__(self, name):
         self.name = name
         self.stats = collections.defaultdict(lambda: 0)
+        self.gauges = {}
```
Comment (matthewdupre) [inline calico/felix/futils.py]: It's a bit strange that a StatCounter is a Gauge...

**Comment by matthewdupre** [comment]
So we're going to ship the 14.04 and EL7 versions of python-prometheus-client ourselves?

**Comment by matthewdupre** [comment]
Looks like we are.  I think you'll also need to update libcalico (the frozen requirements .txt).

@tomdee might know of more places: for adding a new python dependency to Felix, is there anywhere else we need to update?

**Comment by tomdee** [comment]
The frozen requirements are generated by running `make update-frozen`

File: calico/felix/futils.py (hunk 1 above)
Comment (fasaxc) [inline calico/felix/futils.py]: I don't think Prometheus have a dedicated counter type.  Gauge has inc/dec methods instead.

**Comment by fasaxc** [comment]
@matthewdupre I added https://github.com/projectcalico/libcalico/pull/114 and updated the comments around the use of Gauge.

**Pull request merged**